Promote Topic and pullsubscription API to v1

danyinggu opened the pull request:
Fixes #1426 
Help with #1409, #1410, #1411, #1412 and #1413.

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

- 🎁 Promote Topic and pullSubscription API to v1.
- 🗑️ Remove `IdentityStatus.ServiceAccountName` (see discussion in #1426) and PullSubscriptionSpec.Mode` (see comments in the PR) in v1.
- 🐛 Fix the logic in Topic.CheckImmutableFields for all versions.
- 🐛 Add more test cases in UTs of topic_validation_test and pullsubscription_validate_test for all versions.
- 🐛 Fix wrong naming in pullsubscription_lifecycle_test for all versions.

<!--
If this change has user-visible impact, follow the instructions below.
Examples include:

- 🎁 Add new feature
- 🐛 Fix bug
- 🧽 Update or clean up current behavior
- 🗑️ Remove feature or internal logic

Otherwise delete the rest of this template.
-->

**Release Note**

<!--
🗒️ If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other Knative contributors. If this
change has no user-visible impact, no release-note is needed.
-->

```release-note
`IdentityStatus.ServiceAccountName` is removed.
`PullSubscriptionSpec.Mode` is removed.

```

**Docs**

<!--
📖 If this change has user-visible impact, link to an issue or PR in
https://github.com/knative/docs.
-->

capri-xiyue commented:
/lgtm
/hold in case @nacho has any other comments.

danyinggu commented:
/retest

nachocano commented on pkg/apis/intevents/v1/pullsubscription_defaults.go:
```diff
@@ -0,0 +1,60 @@
+	"knative.dev/pkg/ptr"
+)
+
+const (
+	defaultRetentionDuration = 7 * 24 * time.Hour
+	defaultAckDeadline       = 30 * time.Second
+)
```
Can we move this constants to pkg/apis/intevents/register.go? Similar to what we did for the duck ones, so that we avoid duplicating this. 
And make v1alpha1, v1beta1 and this use those?

nachocano commented on pkg/apis/intevents/v1/pullsubscription_defaults.go:
```diff
@@ -0,0 +1,60 @@
+
+	ss.PubSubSpec.SetPubSubDefaults(ctx)
+
+	switch ss.Mode {
+	case ModeCloudEventsBinary, ModeCloudEventsStructured, ModePushCompatible:
+		// Valid Mode.
+	default:
+		// Default is CloudEvents Binary Mode.
+		ss.Mode = ModeCloudEventsBinary
+	}
```
I think we should actually remove Mode from PullSubscription...
If you want to do it in a follow up, that's fine, just create an issue so that we tackle this just after this one gets merged

nachocano replied:
given the flakiness that exists in the CI, it might be better doing it now actually

nachocano commented on pkg/apis/intevents/v1/pullsubscription_lifecycle_test.go:
```diff
@@ -0,0 +1,508 @@
+	}
+)
+
+func TestPubSubStatusIsReady(t *testing.T) {
```
nit: TestPullSubscriptionStatusIsReady..

probably is wrong in the other versions as well, can you update it?

danyinggu replied:
All versions are wrong with the naming... I will update all of them. Thanks for the comment!

nachocano commented on pkg/apis/intevents/v1/pullsubscription_lifecycle_test.go:
```diff
@@ -0,0 +1,508 @@
+	}
+}
+
+func TestPubSubStatusGetCondition(t *testing.T) {
```
same here

nachocano commented on pkg/apis/intevents/v1/pullsubscription_types.go:
```diff
@@ -0,0 +1,230 @@
+	// +optional
+	Transformer *duckv1.Destination `json:"transformer,omitempty"`
+
+	// Mode defines the encoding and structure of the payload of when the
+	// PullSubscription invokes the sink.
+	// +optional
+	Mode ModeType `json:"mode,omitempty"`
```
let's ge rid of this. We are always using Binary mode, so no need for this...

nachocano commented on pkg/apis/intevents/v1/pullsubscription_types.go:
```diff
@@ -0,0 +1,230 @@
+	return defaultRetentionDuration
+}
+
+type ModeType string
+
+const (
+	// ModeCloudEventsBinary will use CloudEvents binary HTTP mode with
+	// flattened Pub/Sub payload.
+	ModeCloudEventsBinary ModeType = "CloudEventsBinary"
+
+	// ModeCloudEventsStructured will use CloudEvents structured HTTP mode with
+	// flattened Pub/Sub payload.
+	ModeCloudEventsStructured ModeType = "CloudEventsStructured"
+
+	// ModePushCompatible will use CloudEvents binary HTTP mode with expanded
+	// Pub/Sub payload that matches how Cloud Pub/Sub delivers a push message.
+	ModePushCompatible ModeType = "PushCompatible"
+)
```
all these will go away

nachocano commented on pkg/apis/intevents/v1/pullsubscription_validation.go:
```diff
@@ -0,0 +1,133 @@
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+)
+
+const (
+	minRetentionDuration = 10 * time.Second   // 10 seconds.
+	maxRetentionDuration = 7 * 24 * time.Hour // 7 days.
+
+	minAckDeadline = 0 * time.Second  // 0 seconds.
+	maxAckDeadline = 10 * time.Minute // 10 minutes.
+)
```
same as the other comment with constants. Can we move them to pkg/apis/intevents/register.go so that we avoid duplicating them for all versions?

nachocano commented on pkg/apis/intevents/v1/pullsubscription_validation.go:
```diff
@@ -0,0 +1,133 @@
+		}
+	}
+
+	// Mode [optional]
+	switch current.Mode {
+	case "", ModeCloudEventsBinary, ModeCloudEventsStructured, ModePushCompatible:
+		// valid
+	default:
+		errs = errs.Also(apis.ErrInvalidValue(current.Mode, "mode"))
+	}
```
this will go away

nachocano commented on pkg/apis/intevents/v1/pullsubscription_validation.go:
```diff
@@ -0,0 +1,133 @@
+	// Modification of Topic, Secret and Project are not allowed. Everything else is mutable.
+	if diff := cmp.Diff(original.Spec, current.Spec,
+		cmpopts.IgnoreFields(PullSubscriptionSpec{},
+			"Sink", "Transformer", "Mode", "AckDeadline", "RetainAckedMessages", "RetentionDuration", "CloudEventOverrides")); diff != "" {
```
will have to remove Mode from here

nachocano commented on pkg/apis/intevents/v1/topic_validation.go:
```diff
@@ -0,0 +1,69 @@
+	var errs *apis.FieldError
+
+	// Topic is immutable.
+	if original.Spec.Topic != current.Spec.Topic {
```
we need to make other attributes immutable as well. Similar to what we had for PullSubscription.

nachocano commented:
Thanks for doing this @danyinggu... I added few comments.
Can you update the PR description as say something `Helps with <Issue number>`, so that we can track it in the issue as well. After the updates, `The logic is exactly the same as v1beta1.` won't apply anymore.

Also, we decided to remove IdentityStatus.ServiceAccountName. I know is not related to this one, but can we add it here as well. My worry is that the CI is so unstable, so once we get a green, I'd rather merge that change as well...

nachocano commented:
@capri-xiyue once @danyinggu does the changes, you might be better off by merging her branch so that you can move forward with the reconcilers, in case the CI remains unstable and this takes a while to merge...

danyinggu commented on pkg/apis/intevents/v1/topic_validation.go:
```diff
@@ -54,16 +56,14 @@ func (current *Topic) CheckImmutableFields(ctx context.Context, original *Topic)
-			})
+	// Modification of Topic, Secret, ServiceAccountName and Project are not allowed. Everything else is mutable.
+	if diff := cmp.Diff(original.Spec, current.Spec,
+		cmpopts.IgnoreFields(TopicSpec{}, "PropagationPolicy", "EnablePublisher")); diff != "" {
```
@nachocano, could you please check the logic is ok here? You mentioned that all fields that are immutable in pullsubscription should also be immutable here, does that mean `PropagationPolicy` and `EnablePublisher` are mutable then?

nachocano replied:
my worry of letting them be mutable is that if we change them, then our reconciler might not properly reconcile those changes. I think for these two it should be ok. We can leave them mutable for now.
Do you mind creating a new issue to check immutability constraints in all of our resources, and whether we are doing it right. We can then tackle that in the following sprint.

danyinggu replied:
Sure, issue created #1448

danyinggu commented on pkg/apis/intevents/v1/topic_validation_test.go:
```diff
@@ -105,6 +115,54 @@ func TestTopicCheckImmutableFields(t *testing.T) {
+			},
+			allowed: false,
+		},
+		"ServiceAccountName changed": {
```
@nachocano Just want to confirm - in sources, we have ServiceAccountName should be immutable, so I think in topic and pullsubscription, it should be the same case. So I added some UT to verify that. Is that okay?

nachocano replied:
yes, that's ok, thanks @danyinggu !

danyinggu commented on pkg/apis/intevents/v1alpha1/pullsubscription_validation_test.go:
```diff
@@ -311,6 +340,23 @@ func TestPubSubCheckImmutableFields(t *testing.T) {
 			},
 			allowed: false,
 		},
+		"ServiceAccountName changed": {
```
@nachocano Same as the topic_validation UT, I found we didn't check ServiceAccount is immutable here, so I added it for all versions.

nachocano replied:
sg, thanks!

danyinggu commented on pkg/apis/intevents/v1alpha1/topic_validation.go:
```diff
@@ -55,16 +57,16 @@ func (current *Topic) CheckImmutableFields(ctx context.Context, original *Topic)
-				Details: fmt.Sprintf("was %q, now %q", original.Spec.Topic, current.Spec.Topic),
-			})
+	if diff := cmp.Diff(original.Spec, current.Spec,
+		cmpopts.IgnoreFields(TopicSpec{}, "PropagationPolicy", "EnablePublisher")); diff != "" {
```
Same as above... I added the topic immutable check logic to all versions to make them consistent.

nachocano replied:
thanks!

danyinggu commented:
> Thanks for doing this @danyinggu... I added few comments.
> Can you update the PR description as say something `Helps with <Issue number>`, so that we can track it in the issue as well. After the updates, `The logic is exactly the same as v1beta1.` won't apply anymore.
> 
> Also, we decided to remove IdentityStatus.ServiceAccountName. I know is not related to this one, but can we add it here as well. My worry is that the CI is so unstable, so once we get a green, I'd rather merge that change as well...

`IdentityStatus.ServiceAccountName` is removed in this PR. I re-editted the PR description and added release note. Also linked the ServiceAccountName issue.

nachocano commented:
/lgtm
/approve

danyinggu commented:
/retest

capri-xiyue commented:
> @capri-xiyue once @danyinggu does the changes, you might be better off by merging her branch so that you can move forward with the reconcilers, in case the CI remains unstable and this takes a while to merge...

I did some investigation. It looks like before I move forward with the reconcilers, I need to make change for the conversion and crd yaml. I will start to make changes for conversion tests for the CloudStorageSource and later work on other resources once Danying gets her pr merged.

capri-xiyue commented:
/test pull-google-knative-gcp-integration-tests

danyinggu commented:
/retest

capri-xiyue commented:
/approved

nachocano commented:
/lgtm
/approve

danyinggu commented:
/retest

danyinggu commented:
/retest

capri-xiyue commented:
/test pull-google-knative-gcp-integration-tests
